File: 5.py
import numpy as np

lines = [e.strip() for e in open(0)]
ranges=[[*map(int,e.split('-'))] for e in lines[:lines.index('')]]
ranges.sort()

# Part 1
a = 0
for ingredient in [int(e) for e in lines[lines.index('')+1:]]:
    for r in ranges:
        if ingredient >= r[0] and ingredient <= r[1]:
            a += 1
            break
print(a)

# Part 2
a = 0
cursor = 0
for r in ranges:
    if cursor < r[0]:
        cursor = r[0]
    if cursor > r[1]:
        continue
    a += r[1]+1 - cursor
    cursor = r[1]+1
print(a)

File: 5.cs
using System.Numerics;

var chunks = Console
    .In.ReadToEnd()
    .ReplaceLineEndings()
    .TrimEnd(Environment.NewLine.ToCharArray())
    .Split(Environment.NewLine + Environment.NewLine);

var ranges = (
    from line in chunks[0].Split(Environment.NewLine)
    let range = line.Split("-").Select(BigInteger.Parse)
    orderby range.First()
    select (Start: range.First(), End: range.Last() + 1)
).ToList();

// Part 1
Console.WriteLine(
    chunks[1]
        .Split(Environment.NewLine)
        .Select(BigInteger.Parse)
        .Count(i => ranges.Any(r => r.Start < i && i <= r.End))
);

// Part 2
BigInteger sum = 0;
BigInteger cursor = 0;
foreach (var (start, end) in ranges)
{
    if (cursor >= end)
        continue;
    sum += end - BigInteger.Max(cursor, start);
    cursor = end;
}
Console.WriteLine(sum);
